prow/git/v2/interactor.go

```diff
@@ -52,3 +62,262 @@ type Interactor interface {
+}
+
+// BranchExists returns true if branch exists in heads.
+func (i *interactor) BranchExists(branch string) bool {
+	i.logger.Infof("Checking if branch %q exists", branch)
+	_, err := i.executor.Run("ls-remote", "--exit-code", "--heads", "origin", branch)
+	return err == nil
+}
+
+// CheckoutNewBranch creates a new branch and checks it out.
+func (i *interactor) CheckoutNewBranch(branch string) error {
+	i.logger.Infof("Checking out new branch %q", branch)
+	if out, err := i.executor.Run("checkout", "-b", branch); err != nil {
+		return fmt.Errorf("error checking out new branch %q: %v %v", branch, err, string(out))
+	}
+	return nil
+}
+
+// Merge attempts to merge commitlike into the current branch. It returns true
+// if the merge completes. It returns an error if the abort fails.
+func (i *interactor) Merge(commitlike string) (bool, error) {
+	return i.MergeWithStrategy(commitlike, "merge")
+}
+
+// MergeWithStrategy attempts to merge commitlike into the current branch given the merge strategy.
+// It returns true if the merge completes. It returns an error if the abort fails.
+func (i *interactor) MergeWithStrategy(commitlike, mergeStrategy string) (bool, error) {
+	i.logger.Infof("Merging %q using the %q strategy", commitlike, mergeStrategy)
+	switch mergeStrategy {
+	case "merge":
+		return i.mergeMerge(commitlike)
+	case "squash":
+		return i.squashMerge(commitlike)
+	default:
+		return false, fmt.Errorf("merge strategy %q is not supported", mergeStrategy)
+	}
+}
+
+func (i *interactor) mergeMerge(commitlike string) (bool, error) {
+	out, err := i.executor.Run("merge", "--no-ff", "--no-stat", "-m", "merge", commitlike)
+	if err == nil {
+		return true, nil
+	}
+	i.logger.WithError(err).Warnf("Error merging %q: %s", commitlike, string(out))
+	if out, err := i.executor.Run("merge", "--abort"); err != nil {
+		return false, fmt.Errorf("error aborting merge of %q: %v %v", commitlike, err, string(out))
+	}
+	return false, nil
+}
+
+func (i *interactor) squashMerge(commitlike string) (bool, error) {
+	out, err := i.executor.Run("merge", "--squash", "--no-stat", commitlike)
+	if err != nil {
+		i.logger.WithError(err).Warnf("Error staging merge for %q: %s", commitlike, string(out))
+		if out, err := i.executor.Run("reset", "--hard", "HEAD"); err != nil {
+			return false, fmt.Errorf("error aborting merge of %q: %v %v", commitlike, err, string(out))
+		}
+		return false, nil
+	}
+	out, err = i.executor.Run("commit", "--no-stat", "-m", "merge")
+	if err != nil {
+		i.logger.WithError(err).Warnf("Error committing merge for %q: %s", commitlike, string(out))
+		if out, err := i.executor.Run("reset", "--hard", "HEAD"); err != nil {
+			return false, fmt.Errorf("error aborting merge of %q: %v %v", commitlike, err, string(out))
+		}
+		return false, nil
+	}
+	return true, nil
+}
+
+// Only the `merge` and `squash` strategies are supported.
+func (i *interactor) MergeAndCheckout(baseSHA string, mergeStrategy string, headSHAs ...string) error {
```

Comment:
Given this checkouts first and merges later, should this be called `CheckoutAndMerge`?

Reply:
This simply achieves parity with the old client. I don't think we should change the way this works in this PR. Follow-ups after the `v1` client is removed could change this.